### PR TITLE
faction camps: improve handle_camp_mission()

### PR DESCRIPTION
**Pull request by mlangsdorf**
#### Summary
```SUMMARY: Infrastructure "faction camps: improve handle_camp_mission()"```

#### Purpose of change
While testing #26764, various minor bugs and issues in the faction camp infrastructure were discovered.  Fixing all of them is too large a patchset, so this PR resolves a bunch of issues with the main
faction camp resolving code, handle_camp_mission()

#### Describe the solution
This is mostly an infrastructure fix with a couple of minor bugfixes.
* define a using for std::vector<std::shared_ptr<npc>> because the full form takes up a huge amount of a line to convey very little information.
* have all the functions that passed an npc * get rewritten to use npc & when possible, and if not, to use shared_ptr<npc> instead.
* mission_key was a vector of mission_entry, and next to it was a string to string map of mission key names and mission descriptions.  There was extra overhead to keep to the two structures in sync. Replace the map by adding the mission description to the mission_entry.
* Added standard functions to mission_key.add for starting and returning missions.  This doesn't really change anything, but it does make it a little easier to follow what's going on.
* the update_time_ functions used to take an npc * from the std::vector<std::shared_ptr<npc>>, and update that npc's return time.  change them to take the vector and update all the NPCs.
* in display_and_choose_opts(), it used to be possible to select missions that were unavailable.  Check that a mission is available, since display_and_choose_opts() has to have that information, before confirming it.  And since only available missions are every sent to handle_camp_missions(), remove the checks in each mission to see if it's available.
* in handle_camp_mission(), when dealing with expansions, loop through all the expansions until finding one that matches the current key's direction, then check that expansion for all possible missions, instead of checking each mission for all possible directions.  Performance wise, I don't really think there's a difference, but it removes a lot of repeated for loops from the code.

#### Describe alternatives you've considered
display_and_choose_opts() should possibly skip over unavailable and grayed out missions, but it's good
enough.

As usual with faction camps, there's a lot more to do and a lot more that could have been done within
this PR, but this is enough for now.

#### Additional context
See also #26920 which is mostly based on this PR.

### CODE REVIEW

File: src/faction_camp.cpp [hunk 1]
```diff
@@ -85,7 +88,7 @@ static std::array<std::pair<const std::string, point>, COMPANION_SORT_POINTS> so
     }
 };
 
-static bool update_time_left( std::string &entry, const std::shared_ptr<npc> comp )
+static bool update_time_left( std::string &entry, const npc_ptr comp )
```
Comment (BevapDin) [inline src/faction_camp.cpp]: A reference to `npc` would be much better than a pointer. Currently the pointer is derferenced without any check for it being null. It is clearly expected to never be null. Using a reference as parameter type makes this clear to the caller.

Same with the other functions below.

Comment (mlangsdorf) [reply src/faction_camp.cpp]: Fixed.

File: src/faction_camp.cpp [hunk 2]
```diff
@@ -149,660 +161,587 @@ void talk_function::camp_missions( mission_data &mission_key, npc &p )
...
-                                          "> fruits and nuts dependening on season\n"
-                                          "May produce less food than consumed!\n"
-                                          "Risk: Very Low\n"
-                                          "Time: 4 Hours, Repeated\n"
-                                          "Positions: %d/3\n" ), npc_list.size() );
-        mission_key.push( "Camp Forage", _( "Forage for plants" ), "", false, npc_list.size() < 3 );
+    if( bcp->has_level( camp_ctr, 2, base_dir ) ) {
+        comp_list npc_list = companion_list( p, "_faction_camp_foraging" );
+        entry = string_format( _( "Notes:\n"
+                                  "Send a companion to edible plans.\n \n"
```
Comment (I-am-Erk) [inline src/faction_camp.cpp]: What kind of edible plans are these? I'm picturing a russian superspy consuming the microfiche before james bond can stop him.

You could change it to `"Send a companion to gather edible plants.\n \n"` but I'd be sad about the loss of james bond references.

Comment (mlangsdorf) [reply src/faction_camp.cpp]: Fixed.

**Comment by kevingranade** [comment]
I merged in the spelling edits that landed recently.

**Pull request merged**